endless/eosprofile.c

```diff
@@ -0,0 +1,580 @@
+  res->file = g_strdup (file);
+  res->line = line;
+
+  res->samples = g_array_sized_new (FALSE, FALSE, sizeof (ProfileSample), N_SAMPLES);
+
+  res->ref_count = 1;
+
+  g_mutex_init (&res->probe_lock);
+
+  return res;
+}
+
+static void
+eos_profile_probe_free (EosProfileProbe *probe)
+{
+  G_LOCK (profile_state);
+  g_hash_table_remove (profile_state->probes, probe->name);
+  G_UNLOCK (profile_state);
+
+  g_array_unref (probe->samples);
+  g_free (probe->name);
+  g_free (probe->function);
+  g_free (probe->file);
+
+  g_free (probe);
+}
+
+static EosProfileProbe *
+eos_profile_probe_ref (EosProfileProbe *probe)
+{
+  if (probe == &eos_profile_dummy_probe)
+    return probe;
+
+  g_atomic_int_inc (&probe->ref_count);
+}
+
+static void
+eos_profile_probe_unref (EosProfileProbe *probe)
+{
+  if (probe == &eos_profile_dummy_probe)
+    return;
+
+  if (g_atomic_int_dec_and_test (&probe->ref_count))
+    eos_profile_probe_free (probe);
+}
+
+G_DEFINE_BOXED_TYPE (EosProfileProbe, eos_profile_probe,
+                     eos_profile_probe_ref,
+                     eos_profile_probe_unref)
+
+/**
+ * eos_profile_probe_start:
+ * @file: the source file for the probe, typically represented by %__FILE__
+ * @line: the line in the source @file, typically represented by %__LINE__
+ * @function: the function for the probe, typically represented by %G_STRFUNC
+ * @name: a unique name for the probe
+ *
+ * Starts a profiling probe for @name, creating it if necessary.
+ *
+ * Returns: (transfer none): a profile probe identifier; use eos_profile_probe_stop()
+ *   to stop the profiling on the returned probe
+ *
+ * Since: 0.6
+ */
+EosProfileProbe *
+eos_profile_probe_start (const char *file,
+                         gsize       line,
+                         const char *function,
+                         const char *name)
+{
+  /* Don't measure the lock */
+  gint64 sample_time = g_get_monotonic_time ();
+
+  /* We can take this out of the lock because by the time we reach
+   * eos_profile_probe_stop() the profile state is guaranteed to
+   * either always exist, or not
+   */
+  if (profile_state == NULL)
+    return &eos_profile_dummy_probe;
+
+  G_LOCK (profile_state);
+
+  EosProfileProbe *res = g_hash_table_lookup (profile_state->probes, name);
+  if (res == NULL)
+    {
+      res = eos_profile_probe_new (file, line, function, name);
+
+      g_hash_table_insert (profile_state->probes, res->name, res);
+    }
+
+  g_array_append_vals (res->samples,
+                       &(ProfileSample) {
+                         .start_time = sample_time,
+                         .end_time = -1
+                       },
+                       1);
+
+  G_UNLOCK (profile_state);
+
+  return (EosProfileProbe *) res;
+}
+
+/**
+ * eos_profile_probe_stop:
+ * @probe: a #EosProfileProbe
+ *
+ * Stops a profiling probe started using eos_profile_probe_start().
+ *
+ * Since: 0.6
+ */
+void
+eos_profile_probe_stop (EosProfileProbe *probe)
```

Comment:
When calling this from JS, my probe is not getting stopped. (That is, the samples' end time is still -1 at the end of execution.)
Here are the probes I added to eos-knowledge-lib in case you want to try to reproduce it:
```diff
diff --git a/configure.ac b/configure.ac
index c72f8631e..308480dbe 100644
--- a/configure.ac
+++ b/configure.ac
@@ -182,6 +182,7 @@ PKG_CHECK_MODULES([EOS_KNOWLEDGE_PRIVATE], [
     gstreamer-audio-1.0 >= 1.10
     gstreamer-video-1.0 >= 1.10
     epoxy
+    endless-0 >= 0.5
 ])
 PKG_CHECK_MODULES([GRESOURCE_PLUGIN], [
     gio-2.0
diff --git a/js/app/modules/selection/xapian.js b/js/app/modules/selection/xapian.js
index 645ebc8e7..ca7f06e5a 100644
--- a/js/app/modules/selection/xapian.js
+++ b/js/app/modules/selection/xapian.js
@@ -3,6 +3,7 @@
 // Copyright 2016 Endless Mobile, Inc.
 
 const Eknc = imports.gi.EosKnowledgeContent;
+const Endless = imports.gi.Endless;
 
 const Module = imports.app.interfaces.module;
 const Selection = imports.app.modules.selection.selection;
@@ -97,6 +98,7 @@ var Xapian = new Module.Class({
 
         this._loading = true;
         this.notify('loading');
+        let probe = Endless.ProfileProbe.start('xapian.js', 100, 'queue_load_more', 'loading more');
         engine.query_promise(query)
         .then(({ models, upper_bound }) => {
             this._loading = false;
@@ -170,6 +172,9 @@ var Xapian = new Module.Class({
                 this._error_state = true;
                 this.notify('in-error-state');
             }
+        })
+        .then(() => {
+            probe.stop();
         });
     },
 
diff --git a/lib/eosknowledgeprivate/ekn-util.c b/lib/eosknowledgeprivate/ekn-util.c
index d6582b1a4..e4c1783ab 100644
--- a/lib/eosknowledgeprivate/ekn-util.c
+++ b/lib/eosknowledgeprivate/ekn-util.c
@@ -1,4 +1,7 @@
 #include "config.h"
+
+#include <endless/endless.h>
+
 #include "ekn-util.h"
 
 /**
@@ -220,6 +223,8 @@ ekn_style_context_get_custom_string (GtkStyleContext *context,
 gchar*
 ekn_extract_pixbuf_dominant_color (GdkPixbuf *pixbuf)
 {
+  g_autoptr(EosProfileProbe) probe = EOS_PROFILE_PROBE("dominant color determination");
+
   const guint8 *pixels, *pixel;
   const gint skip_x = 10, skip_y = 10, alpha_threshold = 40;
   const gdouble rgb_range = 255.0, hue_range = 359.0, sv_threshold = 0.3;
```

Reply:
Indeed, the issue was the `GBoxed` implementation freeing the probe, instead of stopping it once it reaches a zero reference count. I've modified the pull request.

Reply:
The boxed type now does not do any type of memory management — `copy()` returns the pointer, and `free()` doesn't do anything.

This reflects the memory management semantics underlying the `EosProfileProbe` C type: `start()` allocates a new probe, if needed, and returns a pointer to it, but the memory is fully managed by eos-sdk; `stop()` only stops the sampling.